nifi-nar-bundles/nifi-aws-bundle/nifi-aws-processors/src/main/java/org/apache/nifi/processors/aws/s3/FetchS3Object.java

```diff
@@ -147,6 +171,11 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
             request = new GetObjectRequest(bucket, key, versionId);
         }
         request.setRequesterPays(requesterPays);
+        if(rangeLength != null) {
```

Comment:
If rangeLength is 0, won't this result in a returned range of 1 byte?

Reply:
It doesn't seem to return a range of 1.  If rangeLength is 0, S3 throws an error about it being an invalid range and the file routes to the Failure queue.  A rangeLength of 1B downloads 1 byte as expected.  Am I missing something?  I'll update the validator for rangeLength to use createDataSizeBoundsValidator(1, Long.MAX_VALUE) so that we don't ever get to the zero length error condition.

Reply:
I'm trying to test from your recent changes on the branch and ran into an issue when I set these property values for FetchS3Object
- Range Start = 1
- Range Length = 1

The processor can't validate with that setting as it produces this validation warning:
- 's3-object-range-start' validated against '1' is invalid because Must be of format <Data Size> <Data Unit> where <Data Size> is a non-negative integer and <Data Unit> is a supported Data Unit, such as: B, KB, MB, GB, TB
- Same validation error for 's3-object-range-length'. 


To clarify my original concern, should require Range Length  be > 0 if Range Start is set to a value.  
Since range requests start at 0 byte, It should be valid to have 
- Range Start = 0
- Range Length = 1
with a result indicating that this is requesting just the first byte from the resource.

Reply:
As a follow up, if I change
- Range Start = 1B
- Range Length = 1B

What currently gets returned is bytes 1 and 2, which is returning 1 more byte then requested.

Reply:
At time of writing, im testing this with a small file in an s3 bucket, and with three different FetchS3Object calls on the same, 

See this image of a flow depicting the three different FetchS3Object processors, which only differ in what range start and range length is requested 
![Screenshot from 2020-10-22 18-02-52](https://user-images.githubusercontent.com/14304023/96913477-3365f500-1493-11eb-9e59-3a18f59e8e31.png)

With the following hex representation of a file that is the source being retrieved ... 
```
0x00000000	5B	2E	53	68	65	6C	6C	43	6C	61	73	73	49	6E	66	6F	[.ShellClassInfo
0x00000010	5D	0D	0A	43	4C	53	49	44	3D	7B	36	34	35	46	46	30	]..CLSID={645FF0
0x00000020	34	30	2D	35	30	38	31	2D	31	30	31	42	2D	39	46	30	40-5081-101B-9F0
0x00000030	38	2D	30	30	41	41	30	30	32	46	39	35	34	45	7D	0D	8-00AA002F954E}.
0x00000040	0A	4C	6F	63	61	6C	69	7A	65	64	52	65	73	6F	75	72	.LocalizedResour
0x00000050	63	65	4E	61	6D	65	3D	40	25	53	79	73	74	65	6D	52	ceName=@%SystemR
0x00000060	6F	6F	74	25	5C	73	79	73	74	65	6D	33	32	5C	73	68	oot%\system32\sh
0x00000070	65	6C	6C	33	32	2E	64	6C	6C	2C	2D	38	39	36	34	0D	ell32.dll,-8964.
0x00000080	0A	.
```

from left to right ...
- start 0, length 0 ... yields 1 byte ... viewing queue in nifi, content in hex mode has   `0x00000000    5B  10    [.`
  - Im not sure whether this should error (returning nothing), or return the entire file?
- start 0, length 1 ... yields 2 bytes .. viewing queue in nifi, content in hex mode has  `0x00000000    5B  2E    [.`
  - I was expecting this to return `5B`
- start 1, length 1 ... yields 2 bytes .. viewing queue in nifi, content in hex mode has  `0x00000000    2E  53    .S`
  - I was expecting this to return `2E`

Reply:
Maybe change 
```
request.setRange(rangeStart, rangeStart + rangeLength);
```
to
```
request.setRange(rangeStart, rangeStart + rangeLength - 1);
```

Reply:
Thanks for your feedback.  I just tested it again against real AWS S3 and you are absolutely right.  This is apparently a bug in a third party product we've been using.  I will update with the -1 tomorrow.

As for the validation errors you mentioned, this is using DataSize validation rather than pure long values, which allows shortcuts such as Range start: 0B, Range Length: 1GB, followed by Range start: 1GB, Range length: 1GB.  So to use plain bytes, it requires adding a 'B' to the end.  It seemed slightly cleaner than requiring someone to calculate out to the exact byte.  Do you have a strong opinion against this syntax?  Or a suggestion for making it clearer in the documentation?  I'd be happy to update it.

Reply:
I just committed changes with the -1 for the range length calculation and changed the validators to require a length of at least 1B if it is specified.  I don't think range length should be required when range start is specified because sometimes we have a need to read to the end of the file from a starting offset, and reading to the end is the default behavior if range length is not specified.  Please let me know your thoughts.

Reply:
Personally I'm ok with using 0B, though I'd defer to others when it comes to standards for that.  If its kept that way, I agree, it should have documentation guidance improved.  I plan to take a look at this in the next 24 hours and will report back.

Reply:
My test results using a file with the following contents as input

```
0x00000000  5B 2E 53 68 65 6C 6C 43 6C 61 73 73 49 6E 66 6F  [.ShellClassInfo
0x00000010  5D 0D 0A 43 4C 53 49 44 3D 7B 36 34 35 46 46 30  ]..CLSID={645FF0
0x00000020  34 30 2D 35 30 38 31 2D 31 30 31 42 2D 39 46 30  40-5081-101B-9F0
0x00000030  38 2D 30 30 41 41 30 30 32 46 39 35 34 45 7D 0D  8-00AA002F954E}.
0x00000040  0A 4C 6F 63 61 6C 69 7A 65 64 52 65 73 6F 75 72  .LocalizedResour
0x00000050  63 65 4E 61 6D 65 3D 40 25 53 79 73 74 65 6D 52  ceName=@%SystemR
0x00000060  6F 6F 74 25 5C 73 79 73 74 65 6D 33 32 5C 73 68  oot%\system32\sh
0x00000070  65 6C 6C 33 32 2E 64 6C 6C 2C 2D 38 39 36 34 0D  ell32.dll,-8964.
0x00000080  0A                                               .
```

| Range Start | Range Length | Result | Notes |
| --- | --- | --- | --- | 
| 0B | 0B | PASS | 0B no longer allowed for length, must be set to empty string/unset, and validation picks this up |
| 0B | 1B | PASS | 1 byte returned, hex viewer shows 5B 10 (ok, see note below) |
| 1B | 1B | PASS | 1 byte returned, hex viewer shows 2E 10 (ok, see note below) |
| 0B | 2B | PASS | 2 bytes returned, hex viewer shows 5B 2E |
| 120B | 120B | PASS | 9 bytes returned, hex viewer showes 6C 2C 2D 38 39 36 34 0D 0A, the final 9 bytes of the file |
| 0B | unset | PASS | 129 bytes returned, hex viewer shows full contents of file |
| 120B | unset | PASS | 9 bytes returned, hex viewer showes 6C 2C 2D 38 39 36 34 0D 0A, the final 9 bytes of the file |

Known Issues
- There is some bug in the hex viewer that shows an extra x10 character when the content is only 1 byte.  This isn't introduced by this capability.

The above testing was done manually, so this is something that could be applied in a unit test to automate